chain/byteexec/byteexec.go

```diff
@@ -0,0 +1,105 @@
+package byteexec
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+)
+
+var (
+	fileMode   = os.FileMode(0744)
+	initMutex  sync.Mutex
+	maximumTry = 10
+)
+
+// Exec is a handle to an executable that can be used to create an exec.Cmd
+// using the Command method. Exec is safe for concurrent use.
+type Exec struct {
+	Filename string
+}
+
+func writeFile(data []byte) (string, error) {
+	filename := fmt.Sprintf("./temp%d", rand.Uint64())
+	file, err := os.OpenFile(filename, os.O_RDWR|os.O_CREATE|os.O_EXCL, fileMode)
+	defer file.Close()
+
+	if err != nil {
+		if !os.IsExist(err) {
+			return "", fmt.Errorf("Unexpected error opening %s: %s", filename, err)
+		}
+		return "", nil
+	}
+
+	_, err = file.Write(data)
+	if err != nil {
+		os.Remove(filename)
+		return "", fmt.Errorf("Unable to write to file at %s: %s", filename, err)
+	}
+	return filename, nil
+}
+
+// New creates a new instace of Exec
+func New(data []byte) (Exec, error) {
```

Comment:
Why do we need to create an Exec instance? rather than just have a function to run directly? Creating an instance means the caller needs to remember the close it? 

I think it's cleaner to just have `RunOnDocker` or `RunOnLocal` function that takes code directly, kinda similar to `owasm`'s `Execute`. Not sure if I miss anything.